test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/PreserveTypesAndAttributesTests.cs

```diff
@@ -0,0 +1,395 @@
+//---------------------------------------------------------------------
+// <copyright file="PreserveTypesAndAttributesTests.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.OData.Edm;
+using Xunit;
+
+namespace Microsoft.OData.Client.Tests.ALinq
+{
+    /// <summary>
+    /// Tests to check original resource type is used when retrieving attributes like OriginalNameAttribute
+    /// even if the IQueryable<T> object is casted to another type.
+    /// </summary>
+    public class PreserveTypesAndAttributesTests
+    {
+        private readonly TestContext ctx;
+
+        public PreserveTypesAndAttributesTests()
+        {
+            this.ctx = new TestContext();
+        }
+
+        private static void AssertEqualRequestUri(IQueryable query1, IQueryable query2)
+        {
+            var dsq1 = query1 as DataServiceQuery;
+            var dsq2 = query2 as DataServiceQuery;
+            Assert.Equal(dsq1.RequestUri, dsq2.RequestUri);
+        }
+
+        [Fact]
+        public void OneLevelCase1()
+        {
+            IQueryable<ICustomer> query1 = ctx.OneLevelCustomers1;
+            query1 = query1.Where(c => c.CorporationName != "").OrderBy(c => c.Id).Take(2);
+
+            IQueryable<OneLevelCustomer1> query2 = ctx.OneLevelCustomers1;
+            query2 = query2.Where(c => c.CorporationName != "").OrderBy(c => c.Id).Take(2);
+
+            AssertEqualRequestUri(query1, query2);
+
+            string resp = "[{\"CustomerID\":\"ALFoKI\",\"CompanyName\":\"Alfreds Futterkiste\",\"ContactName\":\"Maria Anders\",\"City\":\"Berlin\",\"Region\":null},{\"CustomerID\":\"ANATR\",\"CompanyName\":\"Ana Trujillo Emparedados y helados\",\"ContactName\":\"Ana Trujillo\",\"City\":\"M\\u00e9xico D.F.\",\"Region\":null}]";
+            ctx.InterceptRequestAndMockResponseValue("Customers", resp);
+
+            var result1 = query1.ToList();
+            var result2 = query2.ToList();
+
+            Assert.Equal(result1.Count, result2.Count);
+            for (var i = 0; i < result1.Count; i++)
+            {
+                Assert.Equal(result1[i].City, result2[i].City);
+                Assert.Equal(result1[i].CorporationName, result2[i].CorporationName);
+                Assert.Equal(result1[i].Name, result2[i].Name);
+                Assert.Equal(result1[i].Id, result2[i].Id);
+            }
+        }
```

Comment:
Given the discussion on `DataServiceContext` and `TestContext` down below, I'd suggest a refactor like the following, getting rid of the `TestContext` in favour of referencing `DataServiceContext` directly:

```suggestion
        {
            ctx.CreateQuery<OneLevelCustomer1`> query =  ctx.CreateQuery<OneLevelCustomer1>("Cusomers");
            IQueryable<ICustomer> queryWithInterface= query;
            queryWithInterface = queryWithInterface.Where(c => c.CorporationName != "").OrderBy(c => c.Id).Take(2);

            IQueryable<OneLevelCustomer1> queryWithClass= query.Where(c => c.CorporationName != "").OrderBy(c => c.Id).Take(2);

            AssertEqualRequestUri(queryWithInterface, queryWithClass);

            string resp = "[{\"CustomerID\":\"ALFoKI\",\"CompanyName\":\"Alfreds Futterkiste\",\"ContactName\":\"Maria Anders\",\"City\":\"Berlin\",\"Region\":null},{\"CustomerID\":\"ANATR\",\"CompanyName\":\"Ana Trujillo Emparedados y helados\",\"ContactName\":\"Ana Trujillo\",\"City\":\"M\\u00e9xico D.F.\",\"Region\":null}]";
            ctx.InterceptRequestAndMockResponseValue("Customers", resp);

            var result1 = queryWithInterface.ToList();
            var result2 = queryWithClass.ToList();

            Assert.Equal(result1.Count, result2.Count);
            for (var i = 0; i < result1.Count; i++)
            {
                Assert.Equal(result1[i].City, result2[i].City);
                Assert.Equal(result1[i].CorporationName, result2[i].CorporationName);
                Assert.Equal(result1[i].Name, result2[i].Name);
                Assert.Equal(result1[i].Id, result2[i].Id);
            }
        }
```

Reply:
In my opinion, this makes the test clearer, reducing the references that need to be looked up, without adding any more clutter to the original version.